### PR TITLE
[dev-launcher][android] Use the `DevSupportManagerFactory`

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Part of ENG-4210.

# How

Uses `ReactNativeHostHandler` to provide custom `DevSupportManagerFactory` if possible. Otherwise, we fall back to the previous solution.

# Test Plan

- run bare-expo and confirm that the dev support manager is created by the factory (rn 67). 
- create a new project with rn 66 and confirm that the dev support manager is swapped by dev-launcher code
- create a new project with rn 67 and with an old version of the `expo-modules-core` and confirm that the dev support manager is swapped by dev-launcher code

### CODE REVIEW

File: packages/expo-dev-launcher/android/src/react-native-67/expo/modules/devlauncher/rncompatibility/DevLauncherDevSupportManagerFactory.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        reactInstanceManagerHelper,
+        packagerPathForJSBundleName,
+        enableOnCreate, redBoxHandler,
+        devBundleDownloadListener,
+        minNumShakes,
+        customPackagerCommandHandlers?.toMutableMap()
+      )
+    }
+  }
+}
```
Comment (lukmccall) [inline packages/expo-dev-launcher/android/src/react-native-67/expo/modules/devlauncher/rncompatibility/DevLauncherDevSupportManagerFactory.kt:39]:
```suggestion
}

```

File: packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/DevLauncherPackage.kt [hunk 2]
```diff
@@ -1,13 +1,20 @@
...
+import expo.modules.core.interfaces.ReactNativeHostHandler
+import expo.modules.devlauncher.rncompatibility.DevLauncherReactNativeHostHandler
 
 class DevLauncherPackage : Package, ReactPackage {
   override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> = DevLauncherPackageDelegate.createNativeModules(reactContext);
 
   override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> = emptyList()
+
+  override fun createReactNativeHostHandlers(context: Context?): List<ReactNativeHostHandler> {
+    return listOf(DevLauncherReactNativeHostHandler())
```
Comment (esamelson) [inline packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/DevLauncherPackage.kt:18]: should this be delegated to the `debug` flavor?

File: packages/expo-dev-launcher/android/src/react-native-67/expo/modules/devlauncher/rncompatibility/DevLauncherDevSupportManagerFactory.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    surfaceDelegateFactory: SurfaceDelegateFactory?
+  ): DevSupportManager {
+    return if (!enableOnCreate) {
+      DisabledDevSupportManager()
+    } else {
+      DevLauncherDevSupportManager(
+        applicationContext,
+        reactInstanceManagerHelper,
+        packagerPathForJSBundleName,
+        enableOnCreate, redBoxHandler,
```
Comment (esamelson) [inline packages/expo-dev-launcher/android/src/react-native-67/expo/modules/devlauncher/rncompatibility/DevLauncherDevSupportManagerFactory.kt:32]:
```suggestion
        enableOnCreate,
        redBoxHandler,
```

**Review by esamelson — APPROVED** [review]
👌

**Review by ajsmth — APPROVED** [review]

**Pull request merged**